onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -272,65 +353,126 @@ static bool MatchPositionEmbeddingSubgraph2(
-      {0, 0, "Gather", {1, 11}, kOnnxDomain}
-  };
-  if (!graph_utils::FindPath(concat_node, true, concat_parent_path, edges, logger)) {
-    DEBUG_LOG("Failed to find path 3.");
-    return false;
-  }
-  // Two paths share the gather node (with second input indices==1)
-  if (edges[1]->GetNode().Index() != gather_node_1.Index()) {
-    DEBUG_LOG(" Gather nodes in path 1 and 3 expected to be same node.");
-    return false;
-  }
-  if (edges[0]->GetNode().GetOutputEdgesCount() != 1) {
-    DEBUG_LOG("Output edge count not expected for nodes in path 3.");
-    return false;
+  subgraph_node_indices.push_back(position_gather_node.Index());
+  return true;
+}
+
+template <typename T>
+bool CheckEmbeddingData(const T* data, int64_t batch_size, int64_t element_count) {
+  // check that all batches has same data.
+  size_t data_length = batch_size * element_count;
+  for (size_t i = element_count; i < data_length; i++) {
+    if (data[i] != data[i % element_count]) {
+      return false;
+    }
   }
+  return true;
+}
 
-  // Check if the two paths of position gather lead to the same input.
-  if (shape_node_0.MutableInputDefs()[0] != input_ids ||
-      shape_node_1.MutableInputDefs()[0] != input_ids) {
-    DEBUG_LOG("The parent of two shape nodes are expected to be input_ids.");
-    return false;
+static NodeArg* ExtractEmbedding(Graph& graph,
+                                 int64_t batch_size,
+                                 int64_t sequence_length,
+                                 int64_t hidden_size,
+                                 const ONNX_NAMESPACE::TensorProto* tensor) {
+  assert(nullptr != tensor);
+  assert(batch_size > 0);
+  assert(sequence_length > 0);
+  assert(hidden_size > 0);
+
+  auto old_initializer = onnxruntime::make_unique<Initializer>(*tensor);
+  auto data_type = tensor->data_type();
+
+  ONNX_NAMESPACE::TensorProto initializer;
+  initializer.set_name(graph.GenerateNodeArgName("position_embeddings"));
+  initializer.add_dims(sequence_length);
+  initializer.add_dims(hidden_size);
+  initializer.set_data_type(data_type);
+  const int64_t element_count = sequence_length * hidden_size;
+
+  if (data_type == ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+    const float* data = old_initializer->data<float>();
+    if (!CheckEmbeddingData(data, batch_size, element_count)) {
+      return nullptr;
+    }
+
+    initializer.set_raw_data(data, element_count * sizeof(float));
+  } else {  // data_type == ONNX_NAMESPACE::TensorProto_DataType_FLOAT16
+    const MLFloat16* data = old_initializer->data<MLFloat16>();
+    if (!CheckEmbeddingData(data, batch_size, element_count)) {
+      return nullptr;
+    }
+
+    initializer.set_raw_data(data, element_count * sizeof(MLFloat16));
   }
 
-  // Do not add the gather node since it has been added in another path.
-  matched_edges.push_back(edges[0]);
-  return true;
+  NodeArg& node_arg = graph_utils::AddInitializer(graph, initializer);
```

Comment:
old_intializer and new initializer share same raw data. Is the shared raw buffer still valid after the old initializer is cleaned up?

Reply:
old_intializer and new one has same life cycle. After graph_utils::AddInitializer, the weight data will be hold by graph.

Reply:
Graph::Resolve calls Graph::CleanUnusedInitializers.

Reply:
I believe that set_raw_data will get a copy of raw data. You can see that TensorProto has also a function called clear_raw_data(). If it does not own the raw data, there is no need to clear it.